fs/span-manager/span_manager.go

```diff
@@ -331,127 +292,143 @@ func (m *SpanManager) getSpanInfo(offsetStart, offsetEnd compression.Offset) *sp
-		r.Close()
-	})
-	return io.NewSectionReader(r, int64(offset), int64(size)), // doing integer type conversion to allow passing offset and size on the reader
-		nil
-}
-
-func (m *SpanManager) verifySpanContents(compressedData []byte, id compression.SpanID) error {
-	actual := digest.FromBytes(compressedData)
-	expected := m.ztoc.CompressionInfo.SpanDigests[id]
-	if actual != expected {
-		return fmt.Errorf("expected %v but got %v: %w", expected, actual, ErrIncorrectSpanDigest)
-	}
-	return nil
-}
-
-// addSpanToCache adds contents of the span to the cache.
-func (m *SpanManager) addSpanToCache(spanID string, contents []byte, opts ...cache.Option) {
-	if w, err := m.cache.Add(spanID, opts...); err == nil {
-		if n, err := w.Write(contents); err != nil || n != len(contents) {
-			w.Abort()
-		} else {
-			w.Commit()
-		}
-		w.Close()
+	// check again after acquiring lock
+	if s.checkState(uncompressed) {
+		return m.getSpanFromCache(s.id, offsetStart, size)
 	}
-}
 
-// resolveSpanFromCache resolves the span (in Fetched/Uncompressed state) from the cache.
-// This method returns the reader for the uncompressed span.
-// For Uncompressed span, directly return the reader from the cache.
-// For Fetched span, get the compressed span from the cache, uncompress it, cache the uncompressed span and
-// returns the reader for the uncompressed span.
-func (m *SpanManager) resolveSpanFromCache(s *span, offsetStart, size compression.Offset) (io.Reader, error) {
-	id := fmt.Sprintf("%d", s.id)
-	state := s.state.Load().(spanState)
-	if state == uncompressed {
-		r, err := m.getSpanFromCache(id, offsetStart, size)
-		if err != nil {
-			return nil, err
-		}
-		return r, nil
-	}
-	if state == fetched {
-		// get the compressed span from the cache
+	// if cached but not uncompressed, uncompress and cache the span content
+	if s.checkState(fetched) {
+		// get compressed span from the cache
 		compressedSize := s.endCompOffset - s.startCompOffset
-		r, err := m.getSpanFromCache(id, 0, compressedSize)
+		r, err := m.getSpanFromCache(s.id, 0, compressedSize)
 		if err != nil {
 			return nil, err
 		}
 
-		// read the compressed span
+		// read compressed span
 		compressedBuf, err := io.ReadAll(r)
 		if err != nil {
 			return nil, err
 		}
 
-		// uncompress the span
+		// uncompress span
 		uncompSpanBuf, err := m.uncompressSpan(s, compressedBuf)
 		if err != nil {
 			return nil, err
 		}
 
-		// cache the uncompressed span
-		m.addSpanToCache(id, uncompSpanBuf, m.cacheOpt...)
-		err = s.setState(uncompressed)
-		if err != nil {
+		// cache uncompressed span
+		m.addSpanToCache(s.id, uncompSpanBuf, m.cacheOpt...)
+		if err := s.setState(uncompressed); err != nil {
 			return nil, err
 		}
 		return bytes.NewReader(uncompSpanBuf[offsetStart : offsetStart+size]), nil
 	}
-	return nil, ErrSpanNotAvailable
+
+	// fetch-uncompress-cache span: span state can only be `unrequested` since
+	// no goroutine will release span state lock in `requested` state
```

Comment:
This is not true. The lock can be released in the `requested` state if requesting the span fails. 

I think this behavior is still correct though - either we're in the `unrequested` state, or we're in the `requested` state, but need to redrive the request.

Reply:
> This is not true. The lock can be released in the requested state if requesting the span fails.

I think the correct behavior should be: if in `requested` state, the lock has to be held by the requesting goroutine; if the goroutine failed to request the span, it should revert the state to `unrequested` (no one is requesting the span now). (the 2nd half is not implemented yet in this PR).

Otherwise, what another goroutine should do if it acquires the lock and the span in `requested` (which means another goroutine is requesting).

> either we're in the unrequested state, or we're in the requested state, but need to redrive the request.

yeah if a goroutine `requested` the span and failed, it should change state back to `unrequested` before releasing the lock.

Reply:
added the `requested -> unrequested` logic when span fetching fails.